Implement static subgraph optimizations

bkvogel opened the pull request:
Starting from the second iteration, replace the largest static subgraphs in a model with corresponding static schedules to improve runtime performance. This requires the user to add `@static_graph` decorator to the chain's `__call__()` method.

Currently, speedups can be achieved from some models such as the ptb example since much of the Python overhead is removed. However, other than using static schedules, other potential optimizations are not implemented yet, some redundant array copy operations are still performed, and some models may require significantly more memory. This PR is intended to provide the basic API, documentation, and functionality; I then plan to address the memory and other optimizations in future PRs.

This feature is not backward compatible with LSTM, GRU links. However, since it is an optional feature, it will not break existing code.

See the documentation for more information about usage, features, limitations, and examples. Note that the MNIST and CIFAR examples are mostly unmodified except for the model definition code. The ptb example was only modified to unroll the RNN inside of a chain.

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+
+    This decorator should be applied to any code that needs to run each
+    iteration. This should ideally only include the code that performs
+    the actual forward and/or backward computations, and not include code
```
backward computations and not include code

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+    This decorator should be applied to any code that needs to run each
+    iteration. This should ideally only include the code that performs
+    the actual forward and/or backward computations, and not include code
+    for initializaing parameters, checking types, etc.. If the user would
```
etc. ~~If the user would like to...~~ To include other code in a static chain's `__call__()` method
to run every iteration, it should also use this
decorator.

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+            must write its results in-place into these arrays. Any
+            output arrays that may be used inside another schedule
+            function must appear in this list.
+    Returns:
```
Add blank line before `Returns:`

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+    create ndarrays that are used by other functions in the static
+    schedule. In these cases, the following contraints specify how
+    such arrays should be passed into and returned from a function
+    that uses this decorator.
```
A short example of recommended usage would help here.

bkvogel replied:
Thanks for the suggestion. I agree, but I am trying to decide if it is better to include it here or to include in the documentation file (static_graph.rst).

bkvogel replied:
The example is in the "Usage" section:
https://github.com/chainer/chainer/pull/4811/files#diff-d27b1f920f51219d696c5a6f8847728fR28

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+
+    It is suggested to have the function return any output arrays in-place
+    into pre-allocated arrays (1. above) when possible since this provides
+    the most flexability to the scheduler to make various computation speed
```
~~various~~

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,197 @@
+    """Perform checks needed for creation of a static schedule.
+
+    Check if `func` supports static graph optimizations. If not, try
+    to automatically wrap it to be compatible.
```
~~automatically~~

What does it do if it fails to wrap it?

bkvogel replied:
Sorry, it was incorrectly worded. It should not fail, but will raise an exception if it does for some reason.

Crissman commented on examples/static_graph_optimizations/cifar/README.md:
```diff
@@ -0,0 +1,26 @@
+# Convolutional neural networks for CIFAR-10 and CIFAR-100 Classification
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

Crissman commented on examples/static_graph_optimizations/mnist/README.md:
```diff
@@ -0,0 +1,7 @@
+# Multi-Layer Perceptron for MNIST Classification
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

Crissman commented on examples/static_graph_optimizations/mnist/train_mnist_custom_loop.py:
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python
+"""Fully-connected neural network example using MNIST dataset
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

Crissman commented on examples/static_graph_optimizations/ptb/README.md:
```diff
@@ -0,0 +1,9 @@
+# Recurrent Net Language Model
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

Crissman commented on examples/static_graph_optimizations/ptb/gentxt.py:
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+"""Example to generate text from a recurrent neural network language model.
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

Crissman commented on examples/static_graph_optimizations/ptb/train_ptb_custom_loop.py:
```diff
@@ -0,0 +1,304 @@
+"""Sample script of recurrent neural network language model.
```
Add `with static subgraph optimization` to make clear how this is different from the other one.

bkvogel commented:
Thank you for the comments. I have made the requested changes. The main documentation file `static_graph.rst` now includes a usage example with code. I have also updated the documentation to reflect the current limitations (redundant copy operations are no longer performed and memory usage has been reduced compared to the previous version). I have also added a note at the top of the documentation page to warn users that since this is an experimental feature, the API might change in future versions.

bkvogel commented:
@Crissman Do you have any additional comments? Thanks.

Crissman commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+            'func' is called.
+        unique_arrays (list of ndarray): The master list of all unique
+            ndarrays that appear in the static schedule.
+        func_name (str): An optional name of the static function.
```
How would this be used?

Crissman commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+
+        Args:
+            out_vars (list of Variable): The (flattened) list of output
+                variables obtained by performing a defein-by-run
```
define-by-run

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,232 @@
+    Note that this would work during the first iteration, but during
+    the next iteration when "func_A" is called, it would allocate and
+    return a new array for `x` leading to "func_B" reading from the
+    stale reference to `x` from the previous iteration. Actually, such
```
Delete from "Actually, such...

bkvogel replied:
ok

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,232 @@
+    Performance notes:
+
+    It is suggested to have the function return any output arrays in-place
+    into pre-allocated arrays (1. above) when possible since this provides
```
this allows the scheduler to make tradeoffs

Crissman commented on chainer/graph_optimizations/static_graph_utilities.py:
```diff
@@ -0,0 +1,232 @@
+    scheduler may delete the arrays in `inputs` once they are no longer
+    needed in an iteration and then reallocate them again in the next
+    iteration just before the function is called. Note, however, that
+    completely static array allocations are of course not possible if
```
~~of course~~

Crissman commented on examples/static_graph_optimizations/cifar/README.md:
```diff
@@ -0,0 +1,26 @@
+
+The code consists of three parts: dataset preparation, network and optimizer definition and learning loop, similar to the MNIST example.
+
+More models may be added in the future, but currently only one model is available that uses the VGG-style network from [here](http://torch.ch/blog/2015/07/30/cifar.html) which is based on the network architecture from the paper from [here](https://arxiv.org/pdf/1409.1556v6.pdf).
```
This uses the VGG-style...

bkvogel replied:
sorry, I edited the corresponding readme of the CIFAR example (not the one in static graph example) by mistake. I'll updated this one, too.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+            'func' is called.
+        unique_arrays (list of ndarray): The master list of all unique
+            ndarrays that appear in the static schedule.
+        func_name (str): An optional name of the static function.
```
I'll add a note in the documentation to describe.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    static graph `A` contains a static sub-graph `B`. Then, only the chain
+    corresponding to `A` should be marked as static and the chain
+    corresponding
+    to `B` should not be marked as static. (this requirement might be relaxed
```
agreed. I'll remove.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    to reuse the same static schedule during a single forward pass because
+    we assume that there is no need to compute gradients and hence no
+    need to ever perform a backward pass during test mode.
+    It is also possible to disable static optimzations while in test mode.
```
I will document.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+        supplying the keyword argument ``enable_double_backprop=True``
+        to this decorator. Note: this feature has not been tested yet.
+
+    Notes:
```
agreed

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    - When `chainer.config.enable_backprop` is enabled, if a backward pass
+    is not performed each iteration, then the user code must call a method
+    `chain.schedule_manager.end_forward()`on the static chain each iteration.
+    We are considering ways to automate this in the future.
```
OK, agreed

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    and memory usage are possible. The most computation-efficient setting
+    can result in static schedules that use significantly more memory
+    than the default define-by-run code. If this occurs, please change to
+    a more memory-efficient setting and try again.
```
I will update to describe.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+
+        Args:
+            out_vars (list of Variable): The (flattened) list of output
+                variables obtained by performing a defein-by-run
```
sorry. I missed this and a few other changes. Will fix and update again.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    In order to maximize backward compatibility with existing code,
+    we chose to make
+    the behavior of a static chain depend on the training mode flag,
+    `chainer.config.train`. If the it is `True`, then a static chain that is
```
fixed

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    to `B` should not be marked as static. (this requirement might be relaxed
+    in the future).
+
+    In order to maximize backward compatibility with existing code,
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    If test mode is active (`chainer.config.train` is `False`) then it
+    is not necessary to inform the chain at the end of each forward pass
+    because in test mode, a static chain always attempts to reuse
+    existing static schedule objects whenever possible. It is acceptable
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    is not necessary to inform the chain at the end of each forward pass
+    because in test mode, a static chain always attempts to reuse
+    existing static schedule objects whenever possible. It is acceptable
+    to reuse the same static schedule during a single forward pass because
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    existing static schedule objects whenever possible. It is acceptable
+    to reuse the same static schedule during a single forward pass because
+    we assume that there is no need to compute gradients and hence no
+    need to ever perform a backward pass during test mode.
```
I removed later half of sentence to make more concise.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    needs to run every iteration. Specifically, this will contain the code
+    inside any functions called that were annotated with the `@static_code`
+    decorator, which will include all Chainer built-in functions, as well as
+    any user-defined functions that use `@static_codd`. Then, starting
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    However, the user must also be careful of the following:
+    - The user is responsible for applying this decorator correctly. The
+    framework
+    does not check that the define-by-run code actual corresponds to a static
```
ok, removed.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    is not performed each iteration, then the user code must call a method
+    `chain.schedule_manager.end_forward()`on the static chain each iteration.
+    We are considering ways to automate this in the future.
+    - When this feature is enabled, various tradeoffs between computation
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    only executed once. The user is responsible for wrapping such
+    "side effect" code inside a function that is decorated with
+    `@static_code` to ensure that it gets added to the static schedule.
+    - This is still an experimental feature and advanced optimizations such
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    `@static_code` to ensure that it gets added to the static schedule.
+    - This is still an experimental feature and advanced optimizations such
+    as kernel fusion and various memory optimizations are not implemented
+    yet. Expect performance to improve as this feature is developed further.
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+
+    Usage:
+
+    The user is responsible for ensuring that this decorator is only appied
```
OK. I also added a note to refer the user to the documentation for more details and examples of correct usage.

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    to define-by-run code that actually corresponds to a static subgraph.
+    For the best results, apply this decorator to the chains that
+    correspond to the largest static subgraphs in the model.
+    It is not necessary (and not allowed) to
```
ok

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+    static graph `A` contains a static sub-graph `B`. Then, only the chain
+    corresponding to `A` should be marked as static and the chain
+    corresponding
+    to `B` should not be marked as static. (this requirement might be relaxed
```
ok, removed

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+
+    In order to maximize backward compatibility with existing code,
+    we chose to make
+    the behavior of a static chain depend on the training mode flag,
```
fixed

bkvogel commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1479 @@
+
+    Note: If either 'chainer.config.enable_backprop' or 'chainer.config.train'
+    is set to 'False', then cached static schedules will be reused when
+    possible to reduce memory usage. We assume that the user will not need
```
ok

kmaehashi commented:
Jenkins, test this please.

ir5 commented on chainer/graph_optimizations/static_graph.py:
```diff
@@ -0,0 +1,1464 @@
+    static schedule code will be executed instead of its define-by-run code.
+
+    However, the user must also be careful of the following:
+    - The user is responsible for applying this decorator correctly. The
```
These bullet points are not shown appropriately when I read this doc in HTML.

bkvogel replied:
thanks. I'll check.

ir5 commented on examples/static_graph_optimizations/mnist/train_mnist.py:
```diff
@@ -0,0 +1,149 @@
+        self.side_effect_counter = 0
+
+    @static_code
+    def example_side_effect(self):
```
Honestly, I had the impression that this code is a little complicated when I first saw this one.
An example of a side effect code is useful, but I think writing a side effect code in a forward computation like this is relatively a rare situation.
I feel that removing the side effect code will make the example simpler than now. It might be good for casual users.
The current example may be put in another place. How do you think about this?

bkvogel replied:
I agree that it would be rare to write side effect code like this. I added the side effect code to the example after other feedback, so I think it would be good to keep an example somewhere. But I also think it could be better to move it to another place or at least mention in the example that such usage may not be common. Other ideas would be to document an example in a "design notes" document or to just include a test case for it.

bkvogel replied:
For now, I have decided to modify the example so that side effect code is not used by default. However, I have added a second MLP chain to the MNIST example that uses side effect code that the user can optionally enable. I have also updated the documentation to mention the usage of side effect code is expected to be rare but an example is included for completeness. Does it seem okay?

bkvogel commented on examples/static_graph_optimizations/ptb/train_ptb_custom_loop.py:
```diff
@@ -0,0 +1,314 @@
+
+This was required because the `LSTM` link used by the ptb example
+is not currently fully compatible with the static subgraph
+optimizations feature. Specifically, it does not yet support
```
fixed

bkvogel commented on examples/static_graph_optimizations/ptb/train_ptb_custom_loop.py:
```diff
@@ -0,0 +1,314 @@
+slices inside of a static chain.
+
+This was required because the `LSTM` link used by the ptb example
+is not currently fully compatible with the static subgraph
```
fixed

bkvogel commented on examples/static_graph_optimizations/ptb/train_ptb_custom_loop.py:
```diff
@@ -0,0 +1,314 @@
+
+This is a modified version of the standard Chainer ptb example that
+includes static subgraph optimizations. It is mostly unchanged
+from the original model that the RNN is unrolled for `bproplen`
```
fixed

bkvogel commented on docs/source/reference/static_graph.rst:
```diff
@@ -0,0 +1,225 @@
+constructed each iteration. Unfortunately, this flexibility also tends to make dynamic graph frameworks
+slower than static graph frameworks. For example, in Chainer there is a performance penalty involved in
+dynamically constructing the graph each iteration, since it involves creating many objects; each function
+call creates a new `FunctionNode` object as well as creating new `Variable` and array memory allocation
```
I am also working on splitting this file into two parts: basic usage information, and detailed design notes.

bkvogel commented:
Thank you for the comments. I have tried to make the request changes and also split the documentation rst files into two parts: basic usage and a detailed design notes document.

bkvogel commented on chainer/function_node.py:
```diff
@@ -254,7 +257,7 @@ def apply(self, inputs):
             self._input_indexes_to_retain = None
             self._output_indexes_to_retain = None
-            outputs = self.forward(in_data)
+            outputs = static_forward_optimizations(self, in_data)
```
OK, will do.

bkvogel commented on examples/static_graph_optimizations/cifar/models/VGG.py:
```diff
@@ -0,0 +1,124 @@
+import chainer
+import chainer.functions as F
+from chainer.graph_optimizations.static_graph import static_graph
```
I'm fine with a shorter module path like `chainer.static_graph` and renaming the module. I'll look into making the change.

bkvogel replied:
I suppose I should also rename/shorten the import for `static_code`.

bkvogel commented:
Thank you for the comments. I have implemented the requested changes and also updated `examples/graph_optimizations/README.md`.

beam2d commented:
LGTM. Jenkins, test this please.

beam2d commented:
@Crissman Could you approve the changes if ok?

beam2d commented:
Thank you!!!
